### DISCUSSION TITLE
Rest API integration - access token payload is empty?

### DISCUSSION THREAD

**Discussion by raghur** [Budibase/budibase#7020] [Help]
So I'm wondering what I'm doing wrong here and hopefully somebody can help

1. I'm running the latest docker build published 5 days ago
1. I have Auth0 configured and am able to login 
2. I need to be able to call my REST API (only form of integration that's allowed). 
3. I added one of our API endpoints, and in the Authorization, selected OAuth token (that I'd configured)... I've also tried explicitly sending Current user like so 
![image](https://user-images.githubusercontent.com/516267/182145399-c503c233-dacd-404a-b8db-61d8424ce268.png)
4. On the web application side A) I see the request has a JSON web token - but when I paste that into jwt.io, the payload portion is empty `<header>..<signature>` -  Why is the payload being stripped?

I see #3827 - but that says merged... and the token does come through - minus the payload... I'd appreciate a few steps with debugging this scenario as I'm not quite sure what to check

**Comment by raghur** [comment]
On a little additional digging, the token sent to the REST API is a different one other than the access token that that the user gets on logging in...

How do I get budibase to send the user's token as is to the REST api that I'm integrating with?

**Reply by deanhannigan** [reply]
Hi @raghur, thanks for getting in touch! 

The token mapped to `{{ Current User.OAuthToken }}` is the access token supplied by the vendor.
Depending on the provider, they may supply an opaque token that doesn't decode in certain circumstances. That may be the case with Auth0 and explains why you can't decode it, but this shouldn't affect the validity of the token.

Quick checks:
- Are you getting any specific response codes or messages from Auth0 when making the request?

- If your are using the authentication method above, can you confirm that the "Auth" option is set to the correct option in your query? For example, does your "OAuth" config appear in the dropdown and is it selected?
![Screenshot 2022-08-08 at 11 11 14](https://user-images.githubusercontent.com/5913006/183400756-c79c63c2-0c6d-405f-947e-a678b7caf9c6.png)

- If you are manually configuring an "Authorization" header in your query, does it take the form:
![Screenshot 2022-08-08 at 11 54 30](https://user-images.githubusercontent.com/5913006/183402479-38adea71-2814-4a36-840e-2149b378f087.png)

- If you have access to the `global-db` and the users in your system, you can retrieve the OAuth access token from your test user and replace the bindings above with that value.

- If you have curl access, you can perform a quick sanity check on the validity of the token. At minimum you should require the authorization header but add in any other elements required in your request.
`curl https://yourapi.com -H "Authorization: Bearer YOUR_TOKEN"`

- From my understanding with Auth0 you can also verify your access token by pointing a similar request to your `user-info` endpoint e.g: https://YOUR_API_ENDPOINT.us.auth0.com/userinfo

**Reply by raghur** [reply]
Hello Dean, thanks for getting back... I couldn't make any progress so have put this eval on hold for now. I will try your suggestions... but the token is definitely generated by Auth0 but is rejected by our API since it expects certain scopes/permissions. Also, the token payload part of the JWT is empty

Couple of questions

1. When the user logs into the budibase app with SSO, is there a way to send _that_ token to the api? That's how I understood the feature to work - but it seems that budibase requests for another token using the configured client id and client secret and sends that. The trouble is that the configuration does not allow for either specifying scopes/audience etc.[I'm no OIDC/oAuth expert by any means - but I can see the budibase app has parsed custom claims we set on Auth0 after login - but the JWT token that our API recieves is a different one with the empty payload (I can tell this because the algorithm in the JWT header is different.. ]

1. Can you provide more guidance on 
> if you have access to the global-db and the users in your system, you can retrieve the OAuth access token from your test user and replace the bindings above with that value.

**Reply by raghur** [reply]
I was able to get into couchdb - the oAuth access token stored has the blank payload (successive ..) but original profile fetch from auth0 has succeeded as the custom claim attributes are present.

![image](https://user-images.githubusercontent.com/516267/183882631-27d947cc-d307-4a61-a7cb-2219375eb10c.png)

**Reply by raghur** [reply]
Trying the curl command with the saved token does return the user profile... so the token does seem valid that Auth0 accepts and returns userinfo...

So the question comes down to - if the token needs to have certain scopes/audience how'd I do that? Basically are tehre any hooks to customize the token retrieval from auth0 (or any OIDC provider)

**Reply by deanhannigan** [reply]
Hi @raghur,

At the minute we don't support custom authentication scopes, but work to include this functionality is beginning shortly.

The issue can be tracked here: https://github.com/Budibase/budibase/issues/7267
